hare/algorithm.go

```diff
@@ -554,18 +548,47 @@ func (proc *consensusProcess) onRoundEnd(ctx context.Context) {
 	}
 }
 
-// advances the state to the next round.
-func (proc *consensusProcess) advanceToNextRound(ctx context.Context) {
+// advanceToNextRound advances the state to the next round.
+// Returns k, the new round number.
+// Will return an error if unable to advance to next round.
+// The consensuProcess should be stopped if an error is returned.
+func (proc *consensusProcess) advanceToNextRound(ctx context.Context) (uint32, error) {
+	k := proc.getK()
+	// Edge cases:
+	// 1. Preround -> just continue
+	if k == preRound {
+		k = proc.addToK(1)
+		return k, nil
+	}
+	// 2. Already in certify round -> error
+	if k == certifyRound {
+		proc.With().Error("consensus process trying to advance past certify round",
+			log.Uint32("current_k", k),
+			proc.instanceID)
+		return k, errors.New("already at certify round")
+	}
+	// 3. Notify round AND consensus has been reached -> goto certify round
 	if proc.completed && proc.currentRound() == notifyRound {
 		proc.setK(certifyRound)
```

Comment:
i always find it unnecessary to conflate k with the state the consensus process is in. what are we saving, an integer?
the code would be more readable if we separate the consensus state and number of rounds. 
say we add `state` variable to consensusProcess to capture the various round from pre to certify, and k simply capture the number of rounds it went through.

changing k here also cause us to lose debugging information on how many rounds we actually went through.

Reply:
I 1000% agree, and will try to eliminate it when I work on the larger refactor.

Reply:
it used to be a cute idea (when you can keep incrementing k) until certify round is introduced. now it's just confusing.